File: localization/quat_iekf/QuatIEKF.m
% right invariant EKF for quaternion

classdef QuatIEKF < handle
    properties
        g
        g_err

        % q
        % B
        % 
        % q_err
        % B_err
        
        P
       
    end

    methods(Static)

        function result = quat_to_rotm(q)
            q0 = q(1);
            qv = q(2:end);

            q_v_x = [0 -qv(3) qv(2);
                     qv(3) 0 -qv(1);
                     -qv(2) qv(1) 0];

            result = (q0.^2 - qv' * qv) * eye(3) + 2 * qv * qv' - 2 * q0 * q_v_x;

        end

        % quat is a 1x4 array representing a quaternion (w x y z]
        function A = A(quat)

            q_0 = quat(1)';
            q_v = quat(2:end)';
            q_v_x = QuatIEKF.wedge(q_v);
            
            A = (q_0.^2 - q_v' * q_v) * eye(3) + 2 * (q_v * q_v') - 2 * q_0 * q_v_x;
        end

        function result = quat_multiply(q, p)
            % disp(q);
            % disp(p);

            q0 = q(1);
            qv = q(2:end);

            p0 = p(1);
            pv = p(2:end);

            % disp("qv");
            % disp(qv);
            % disp("pv");
            % disp(pv);

            scalar_part = q0 * p0 - qv' * pv;
            vector_part = q0 * pv + p0 * qv + cross(qv, pv);

            result = [scalar_part; vector_part];

        end



        function result = quat_conjugate(q)

            result = [q(1); -1 * q(2:end)];

        end

        % skew symmetric matrix of v
        function result = sup_x(v)
            result = [0 -v(3) v(2);
                   v(3) 0 -v(1);
                   -v(2) v(1) 0];
        end

        
        % see eq 4 in paper
        function result = exp_q(x)
            scalar_part = cos(norm(x));
            vector_part = (x / norm(x)) * sin(norm(x));

            result = [scalar_part; vector_part];
            

        end

        
    end


    methods
        % Q-IEKF initial values
        function obj = QuatIEKF()

            % obj.g = [1 0 0 0 0 0 0]';
            obj.g = [0.0050038, 0.07056, 0.07056, 0.9949962, 0, 0, 0]';

            % see eq 14 in paper
            obj.g_err = [1 0 0 0 0 0 0]';
            obj.P = 500 * eye(6);
           

        end

        % w is a 3x1 array of angular velocities (rad/s)
        % n is a 3x1 array of noise
        function gyro_predict(obj, w, n_v, n_u, dt)
            

            q = obj.g(1:4);
            B = obj.g(5:7);
            q_err = obj.g_err(1:4);
            B_err = obj.g_err(5:7);

            % see eq 21 and 22 of paper
            n_v_est = QuatIEKF.quat_multiply(QuatIEKF.quat_multiply(q, [0; n_v]), QuatIEKF.quat_conjugate(q));
            I_w_est = QuatIEKF.quat_multiply(QuatIEKF.quat_multiply(q, [0; w - B]), QuatIEKF.quat_conjugate(q));
            n_u_est = QuatIEKF.quat_multiply(QuatIEKF.quat_multiply(q, [0; n_u]), QuatIEKF.quat_conjugate(q));
            % disp("n_u:")
            % disp(n_u);
            % disp("I_w_est:");
            % disp(I_w_est);
            % disp("n_u_est");
            % disp(n_u_est);

      
            obj.g_err(1:4) = q_err + QuatIEKF.quat_multiply(-0.5 * q_err, [0; B_err]) * dt + QuatIEKF.quat_multiply(0.5 * n_v_est, q_err) * dt;
            obj.g_err(1:4) = quatnormalize(obj.g_err(1:4)')';
            obj.g_err(5:7) = B_err + cross(QuatIEKF.quat_to_rotm(q_err) * (I_w_est(2:end) - n_u_est(2:end)), B_err) * dt - QuatIEKF.quat_to_rotm(q_err) * n_u_est(2:end) * dt;


            % see eq 27 of paper
            F = [zeros(3), -eye(3);
                 zeros(3), QuatIEKF.sup_x(I_w_est(2:end))];
            G = [QuatIEKF.quat_to_rotm(q)', zeros(3);
                 zeros(3), -QuatIEKF.quat_to_rotm(q)'];
            Q = diag([n_v; n_u].^2);

            % disp("Q");
            % disp(Q);
            % 
            % disp("GQG:");
            % disp(G * (expm(F * dt) * Q * dt * expm(F * dt)') * G');


            % see https://www.youtube.com/watch?v=WHBgSqRPpu4, eq 33 of
            % paper
            obj.P = expm(F * dt) * obj.P * expm(F * dt)' + G * (expm(F * dt) * Q * dt * expm(F * dt)') * G';

            % see https://stackoverflow.com/questions/46908345/integrate-angular-velocity-as-quaternion-rotation
            obj.g(1:4) = QuatIEKF.quat_multiply(q, QuatIEKF.exp_q((w - B) * dt));
            obj.g(5:7) = B;

            % disp("propogation amount:");
            % disp(QuatIEKF.exp_q((w - B) * dt));

            % disp("quat after gyro:");
            % disp(obj.g(1:4));



           

            % obj.P = expm(F * dt) * obj.P * (expm(F * dt))' + expm(G * dt) * Q * (expm(G * dt))'; % TODO: check this error propogation
        end

        % takes in acceleration vector and associated covariance
        function accel_correct(obj, a, cov_a)

            % see https://matthewhampsey.github.io/blog/2020/07/18/mekf,
            % and eq 31, 34 of paper
            H = [QuatIEKF.sup_x([0; 0; -1]), zeros(3)];
            R = QuatIEKF.quat_to_rotm(obj.g(1:4))' * cov_a * QuatIEKF.quat_to_rotm(obj.g(1:4));

           
            % disp("R:");
            % disp(R);

            % disp("obj.P * H'");
            % disp(obj.P * H');
            % 
            % disp()

             % see eq 33 of paper
            K = obj.P * H' / (H * obj.P * H' + R);

            % disp("P:");
            % disp(obj.P);
            % 
            % disp("PH'");
            % disp(obj.P * H');
            % disp("H * obj.P");
            % disp(H * obj.P);
            % disp("HPH'");
            % disp(H * obj.P * H');
            % disp("HPH' + R");
            % disp(H * obj.P * H' + R);
            % disp("R:");
            % disp(R);
            % 
            % disp("K:");
            % disp(K);

            % innovation, see eq 29 of paper
            observation = QuatIEKF.quat_multiply(QuatIEKF.quat_multiply(obj.g(1:4), [0; a]), QuatIEKF.quat_conjugate(obj.g(1:4)));
            predicted = [0; 0; -1];
            innov = predicted - observation(2:end);

            % disp("predicted");
            % disp(predicted);
            % disp("observation:");
            % disp(observation);
            % disp("innovation");
            % disp(innov);
            

            % disp("innov:");
            % disp(innov);

            % see eq 17 of paper
            correction = K * innov;
            
            % disp("correction:")
            % disp(correction);

            

            % update, see eq 20 of paper
            obj.g(1:4) = QuatIEKF.quat_multiply(QuatIEKF.exp_q(-0.5 * correction(1:3)), obj.g(1:4));
            B_update = QuatIEKF.quat_multiply(QuatIEKF.quat_multiply(QuatIEKF.quat_conjugate(obj.g(1:4)), [0; correction(4:6)]), obj.g(1:4));
            % disp("B update:");
            % disp(B_update);
            % 
            obj.g(5:7) = obj.g(5:7) - B_update(2:end);

            % see eq 33 of paper
            obj.P = (eye(6) - K * H) * obj.P;

            % disp("P:");
            % disp(obj.P);


        end

        

        
        
        

      

    end

end

File: localization/quat_iekf/quat_iekf.cpp
#include "quat_iekf.hpp"

namespace mrover {

    QuatIEKF::QuatIEKF() : Node("quat_iekf") {

        q.setIdentity();
        B.setZero();
        q_err.setIdentity();
        B_err.setZero();
        P.setIdentity();

        imu_sub = this->create_subscription<sensor_msgs::msg::Imu>("/imu/data", 10, [&](const sensor_msgs::msg::Imu::ConstSharedPtr& imu_msg) {
            imu_callback(*imu_msg);
        });

        mag_heading_sub = this->create_subscription<mrover::msg::Heading>("/imu/mag", 10, [&](const mrover::msg::Heading::ConstSharedPtr& mag_heading_msg) {
            mag_heading_callback(*mag_heading_msg);
        });


    }

    auto QuatIEKF::quat_to_vec(const Quaterniond& q) -> Vector4d {
        return Vector4d{q.w(), q.x(), q.y(), q.z()};
    }

    auto QuatIEKF::lift_q(const Vector3d& v) -> Quaterniond {

        double scalar_comp = std::cos(v.norm() / 2);
        Vector3d vector_comp = v.normalized() * std::sin(v.norm() / 2);
        return Quaterniond{scalar_comp, vector_comp(0), vector_comp(1), vector_comp(2)};

    }


    void QuatIEKF::predict(const Vector3d& w, const Vector3d& n_v, const Vector3d& n_u, double dt) {

        Vector3d w_meas = w - B;

        Quaterniond n_v_est = q * Quaterniond{0, n_v(0), n_v(1), n_v(2)} * q.conjugate();
        Quaterniond I_w_est = q * Quaterniond{0, w_meas(0), w_meas(1), w_meas(2)} * q.conjugate();
        Quaterniond n_u_est = q * Quaterniond{0, n_u(0), n_u(1), n_u(2)} * q.conjugate();

        // propagate error terms
        Vector4d q_err_vec = quat_to_vec(q_err);
        
        q_err_vec = q_err_vec + quat_to_vec(Quaterniond(-0.5 * q_err_vec) * Quaterniond(0, B_err.x(), B_err.y(), B_err.z())) * dt + 
                                quat_to_vec(Quaterniond(0.5 * quat_to_vec(n_v_est)) * q_err) * dt;

        Vector3d B_err_intermediary = q_err.toRotationMatrix() * (I_w_est.vec() - n_u_est.vec());

        B_err = B_err + B_err_intermediary.cross(B_err) * dt - q_err.toRotationMatrix() * n_u_est.vec() * dt;

        q_err = Quaterniond(q_err_vec);
        q_err.normalize();

        Matrix66d F;
        Matrix66d G;
        Matrix66d Q;
        Vector6d N_vec;

        F << Matrix33d::Zero(), -Matrix33d::Identity(), Matrix33d::Zero(), manif::skew(I_w_est.vec());
        G << q.toRotationMatrix().transpose(), Matrix33d::Zero(), Matrix33d::Zero(), -1 * (q.toRotationMatrix().transpose());
        N_vec << n_v, n_u;
        Q = N_vec.asDiagonal();

        Matrix66d F_d = (F * dt).exp();

        // propagate state, error covariance matrix
        P = F_d * P * F_d.transpose() + G * F_d * Q * dt * F_d.transpose() * G.transpose();
        q = lift_q(w_meas * dt) * q; // see https://math.stackexchange.com/questions/39553/how-do-i-apply-an-angular-velocity-vector3-to-a-unit-quaternion-orientation
        q.normalize();

        RCLCPP_INFO(get_logger(), "Quat: %f, %f, %f, %f", q.x(), q.y(), q.z(), q.w());
        Matrix33d rot_m = q.toRotationMatrix();
        RCLCPP_INFO(get_logger(), "rot_m:\n%f, %f, %f\n%f, %f, %f\n %f, %f, %f", rot_m(0,0), rot_m(0,1), rot_m(0,2),
                                                                                 rot_m(1,0), rot_m(1,1), rot_m(1,2),
                                                                                 rot_m(2,0), rot_m(2,1), rot_m(2,2));

    }

    void QuatIEKF::correct(const Matrix36d& H, const Matrix33d& R, const Vector3d& observation, const Vector3d& predicted) {

        Vector3d innovation = predicted - observation;
        Matrix63d K = P * H.transpose() * (H * P * H.transpose() + R).inverse();
        Vector6d correction = K * innovation;

        RCLCPP_INFO(get_logger(), "innov: %f, %f, %f", innovation(0), innovation(1), innovation(2));
        RCLCPP_INFO(get_logger(), "correction: %f, %f, %f", correction(0), correction(1), correction(2));

        q = lift_q(Vector3d{-correction(0), -correction(1), -correction(2)}) * q;
        q.normalize();
        Quaterniond B_update = q.conjugate() * Quaterniond{0, correction(3), correction(4), correction(5)} * q;
        B = B - Vector3d{B_update.x(), B_update.y(), B_update.z()};
        P = (Matrix66d::Identity() - K * H) * P;

    }

    void QuatIEKF::imu_callback(const sensor_msgs::msg::Imu& imu_msg) {

        Vector3d w{imu_msg.angular_velocity.x, imu_msg.angular_velocity.y, imu_msg.angular_velocity.z};
        Vector3d n_v{imu_msg.angular_velocity_covariance[0], imu_msg.angular_velocity_covariance[4], imu_msg.angular_velocity_covariance[8]};
        Vector3d n_u{0, 0, 0};

        geometry_msgs::msg::Vector3 a = imu_msg.linear_acceleration;
        Matrix33d cov_a;
        cov_a << imu_msg.linear_acceleration_covariance[0], imu_msg.linear_acceleration_covariance[1], imu_msg.linear_acceleration_covariance[2],
                 imu_msg.linear_acceleration_covariance[3], imu_msg.linear_acceleration_covariance[4], imu_msg.linear_acceleration_covariance[5],
                 imu_msg.linear_acceleration_covariance[6], imu_msg.linear_acceleration_covariance[7], imu_msg.linear_acceleration_covariance[8];
        
        double dt = IMU_DT;
        if (last_imu_time) {
            dt = (imu_msg.header.stamp.sec + imu_msg.header.stamp.nanosec * 1e-9) - (last_imu_time.value().sec + last_imu_time.value().nanosec * 1e-9);   
        }

        predict(w, n_v, n_u, dt);

        accel_callback(a, cov_a);

        last_imu_time = imu_msg.header.stamp;

    }

    void QuatIEKF::mag_heading_callback(const mrover::msg::Heading& mag_heading_msg) {

        Matrix36d H;
        Matrix33d R;
        Vector3d observation;
        Vector3d predicted;

        double heading = mag_heading_msg.heading;
        Vector3d mag_ref{1, 0, 0};
        Quaterniond observation_q = q * Quaterniond{0, std::cos(heading), -std::sin(heading), 0} * q.conjugate();
        // Quaterniond observation_q = q * Quaterniond{0, std::cos(heading), std::sin(heading), 0} * q.conjugate();
        RCLCPP_INFO(get_logger(), "mag observation: %f, %f, %f", observation_q.x(), observation_q.y(), observation_q.z());

        H << manif::skew(mag_ref), Matrix33d::Zero();
        R << q.toRotationMatrix().transpose() * mag_heading_msg.heading_accuracy * Matrix33d::Identity() * q.toRotationMatrix();
        observation << observation_q.x(), observation_q.y(), observation_q.z();
        predicted << mag_ref;

        RCLCPP_INFO(get_logger(), "heading: %f", heading);
        correct(H, R, observation, predicted);

    }

    void QuatIEKF::accel_callback(const geometry_msgs::msg::Vector3& accel_msg, const Matrix33d& cov_a) {

        Matrix36d H;
        Matrix33d R;
        Vector3d observation;
        Vector3d predicted;

        Vector3d accel_meas{accel_msg.x, accel_msg.y, -9.81 + accel_msg.z};
        accel_meas.normalize();

        Vector3d accel_ref{0, 0, -1};
        Quaterniond observation_q = q * Quaterniond{0, accel_meas(0), accel_meas(1), accel_meas(2)} * q.conjugate();
        RCLCPP_INFO(get_logger(), "accel observation: %f, %f, %f", observation_q.x(), observation_q.y(), observation_q.z());

        H << manif::skew(accel_ref), Matrix33d::Zero();
        R << q.toRotationMatrix().transpose() * cov_a * q.toRotationMatrix();
        observation << observation_q.x(), observation_q.y(), observation_q.z();
        predicted << accel_ref;

        correct(H, R, observation, predicted);
        
        


    }



};

int main(int argc, char**argv) {

    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<mrover::QuatIEKF>());
    rclcpp::shutdown();
    return 0;

}